src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.ModuleSymbolKey.cs

```diff
@@ -24,12 +26,27 @@ public sealed override void Create(IModuleSymbol symbol, SymbolKeyWriter visitor
                 return default;
             }
 
-            using var result = PooledArrayBuilder<IModuleSymbol>.GetInstance();
-            foreach (var assembly in containingSymbolResolution.OfType<IAssemblySymbol>())
+            using var result = PooledArrayBuilder<IModuleSymbol>.GetInstance(containingSymbolResolution.SymbolCount);
+            foreach (var symbol in containingSymbolResolution)
             {
+                if (symbol is not IAssemblySymbol assembly)
+                    continue;
+
                 // Don't check ModuleIds for equality because in practice, no-one uses them,
                 // and there is no way to set netmodule name programmatically using Roslyn
-                result.AddValuesIfNotNull(assembly.Modules);
+                var assemblyModules = assembly.Modules;
+                if (assemblyModules is ImmutableArray<IModuleSymbol> modules)
+                {
+                    // Avoid allocations if possible
+                    // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2136177
+                    result.AddValuesIfNotNull(modules);
+                }
+                else
+                {
+                    // Visual Basic implementation of IAssemblySymbol.Modules relies on covariance of IEnumerable<T>, so
+                    // the preceding concrete type check will fail.
```

Comment:
We should be able to adjust VB's explicit implementation to use `ImmutableArray.CastUp` to avoid this path.

Reply:
This may be true, but the code path would likely stay in place (just the comment would change). I haven't see VB show up in the perf trace that led to this change so not sure how important this is.

Reply:
> but the code path would likely stay in place (just the comment would change)

Sure, not disagreeing with that. I think I'd prefer if we made this change for VB.

Reply:
➡️ Updated to use `CastUp`